internal-dns-client/src/multiclient.rs

```diff
@@ -0,0 +1,651 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+use crate::types::{DnsKv, DnsRecord, DnsRecordKey, Srv};
+use futures::stream::{self, StreamExt, TryStreamExt};
+use omicron_common::address::{
+    Ipv6Subnet, ReservedRackSubnet, AZ_PREFIX, DNS_PORT, DNS_SERVER_PORT,
+};
+use slog::{info, Logger};
+use std::net::{IpAddr, Ipv6Addr, SocketAddr, SocketAddrV6};
+use trust_dns_resolver::config::{
+    NameServerConfig, Protocol, ResolverConfig, ResolverOpts,
+};
+use trust_dns_resolver::TokioAsyncResolver;
+
+pub type DnsError = crate::Error<crate::types::Error>;
+
+/// Describes how to find the DNS servers.
+///
+/// In production code, this is nearly always [`Ipv6Subnet<AZ_PREFIX>`],
+/// but it allows a point of dependency-injection for tests to supply their
+/// own address lookups.
+pub trait DnsAddressLookup {
+    fn dropshot_server_addrs(&self) -> Vec<SocketAddr>;
+
+    fn dns_server_addrs(&self) -> Vec<SocketAddr>;
+}
+
+fn subnet_to_ips(
+    subnet: Ipv6Subnet<AZ_PREFIX>,
+) -> impl Iterator<Item = IpAddr> {
+    ReservedRackSubnet::new(subnet)
+        .get_dns_subnets()
+        .into_iter()
+        .map(|dns_subnet| IpAddr::V6(dns_subnet.dns_address().ip()))
+}
+
+impl DnsAddressLookup for Ipv6Subnet<AZ_PREFIX> {
+    fn dropshot_server_addrs(&self) -> Vec<SocketAddr> {
+        subnet_to_ips(*self)
+            .map(|address| SocketAddr::new(address, DNS_SERVER_PORT))
+            .collect()
+    }
+
+    fn dns_server_addrs(&self) -> Vec<SocketAddr> {
+        subnet_to_ips(*self)
+            .map(|address| SocketAddr::new(address, DNS_PORT))
+            .collect()
+    }
+}
+
+/// Describes a service which may be inserted into DNS records.
+pub trait Service {
+    fn aaaa(&self) -> crate::names::AAAA;
+    fn srv(&self) -> crate::names::SRV;
+    fn address(&self) -> SocketAddrV6;
+}
+
+/// A connection used to update multiple DNS servers.
+pub struct Updater {
+    log: Logger,
+    clients: Vec<crate::Client>,
+}
+
+impl Updater {
+    pub fn new(address_getter: &impl DnsAddressLookup, log: Logger) -> Self {
+        let addrs = address_getter.dropshot_server_addrs();
+        Self::new_from_addrs(addrs, log)
+    }
+
+    fn new_from_addrs(addrs: Vec<SocketAddr>, log: Logger) -> Self {
+        let clients = addrs
+            .into_iter()
+            .map(|addr| {
+                info!(log, "Adding DNS server: {}", addr);
+                crate::Client::new(&format!("http://{}", addr), log.clone())
+            })
+            .collect::<Vec<_>>();
+
+        Self { log, clients }
+    }
+
+    /// Inserts all service records into the DNS server.
+    ///
+    /// This method is most efficient when records are sorted by SRV key.
+    pub async fn insert_dns_records(
+        &self,
+        records: &Vec<impl Service>,
```

Comment:
I'm not sure exactly how to express this, but I can't help but feeling this should be a map. In the case where `records` is sorted by SRV, then that's effectively what we're doing inside by coalescing all AAAA records and addresses. A map might be frustrating for consumers, though, since it seems like we'll want to `impl Service` for some types (which returns exactly one AAAA and address) and pass one or more of them.

This isn't a correctness issue, I think, so deferring it is fine. I think the `while let` loop on L100 could use a quick comment that we are in fact coalescing, mostly for efficiency in terms of actually inserting the KV records. (I think...)

Reply:
I can do a `&HashMap<SRV, Vec<(AAAA, SocketAddr)>>` - that seems to accomplish what you're describing.

Although I do think this is slightly more burden on consumers, I also think it's more clear on both sides. I've updated to use this format (one SRV, many AAAA/addrs).